napari/layers/utils/layer_utils.py

```diff
@@ -684,58 +684,201 @@ def features_to_pandas_dataframe(features: Any) -> pd.DataFrame:
     return features
 
 
-def _features_from_layer(
-    *,
-    features: Optional[Union[Dict[str, np.ndarray], pd.DataFrame]] = None,
-    properties: Optional[Union[Dict[str, np.ndarray], pd.DataFrame]] = None,
-    property_choices: Optional[Dict[str, np.ndarray]] = None,
-    num_data: Optional[int] = None,
-) -> Tuple[pd.DataFrame, pd.DataFrame]:
-    """Coerces a layer's keyword arguments to feature values and defaults tables.
+class _FeatureManager:
+    """Manages feature values and their defaults.
 
     Parameters
     ----------
-    features : Optional[Union[Dict[str, np.ndarray], pd.DataFrame]]
-        The features input to a layer.
-    properties : Optional[Union[Dict[str, np.ndarray], pd.DataFrame]]
-        The properties input to a layer.
-    property_choices : Optional[Dict[str, np.ndarray]]
-        The property choices input to a layer.
+    values : Optional[Union[Dict[str, np.ndarray], pd.DataFrame]]
+        The features values, which will be passed to the pandas DataFrame initializer.
+        If this is a pandas DataFrame with a non-default index, that index
+        (except its length) will be ignored.
     num_data : Optional[int]
         The number of the elements in the layer calling this, such as
-        the number of points.
-
-    Returns
-    -------
-    pd.DataFrame
-        The pandas DataFrame created from the input features or properties.
-    pd.DataFrame
-        The pandas DataFrame of default values. The names and dtypes will be
-        the same as the features data, but the table will have exactly one row.
-
-    Raises
-    ------
-    ValueError
-        If the input properties columns are not all the same length, or if
-        that length is not equal to the given num_data.
+        the number of points, which is used to check that the features
+        table has the expected number of rows. If None, then the default
+        DataFrame index is used.
     """
-    if properties is not None or property_choices is not None:
-        features = _features_from_properties(
-            properties=properties,
-            property_choices=property_choices,
-            num_data=num_data,
+
+    def __init__(
+        self,
+        values: Optional[Union[Dict[str, np.ndarray], pd.DataFrame]] = None,
```

Comment:
Is there that many places that construct a FeatureManager without  values ? 
Would it be simpler to just say that when contructed it should at least get an empty dict as parameter and avoid those "optional" everywhere ?

Reply:
That's the default behavior of layers that have the `features` parameter, because the default value of that parameter is `None`, so it is common.

In general, I avoid using mutable default parameter values like `{}` to prevent surprising behavior: https://vald-phoenix.github.io/pylint-errors/plerr/errors/basic/W0102.html

If we make this non-optional, but want to avoid that, then we'd need to do something like

```python
if features is None:
  features = {}
```

for every method that takes a `features` parameter.

So, handling it here seems simpler to me.